crates/pathfinder/src/update.rs

```diff
@@ -0,0 +1,170 @@
+        tokio::time::sleep(std::time::Duration::from_secs(60 * 10)).await;
+    }
+}
+
+#[derive(Debug)]
+struct Release {
+    version: String,
+    /// Optional because its possible to have an invalid string ETag header.
+    etag: Option<String>,
+}
+
+#[derive(Debug)]
+enum UpdateResult {
+    /// A new latest release.
+    Update(Release),
+    /// No new releases since last query (status code 304).
+    NotModified,
+    RewestError(reqwest::Error),
+    Other(anyhow::Error),
+}
+
+/// Fetches the latest pathfinder [Release] from Github using their REST API.
+///
+/// The [IF_NONE_MATCH](reqwest::header::IF_NONE_MATCH) header options is set to
+/// the `etag` parameter to prevent Github from sending redundant information. The
+/// resulting 304 status code is mapped to [UpdateResult::NotModified].
+async fn fetch_latest_github_release(
+    client: &reqwest::Client,
+    etag: &Option<String>,
+) -> UpdateResult {
+    use reqwest::StatusCode;
+    use reqwest::Url;
+
+    const USER_AGENT: &str = concat!(
+        env!("CARGO_PKG_NAME"),
+        "/",
+        env!("VERGEN_GIT_SEMVER_LIGHTWEIGHT")
+    );
+    let url = Url::parse("https://api.github.com/repos/eqlabs/pathfinder/releases/latest").unwrap();
+
+    let mut request = client
+        .get(url)
+        .timeout(std::time::Duration::from_secs(300))
+        // https://docs.github.com/en/rest/overview/resources-in-the-rest-api#current-version
+        .header(reqwest::header::ACCEPT, "application/vnd.github.v3+json")
+        // https://docs.github.com/en/rest/overview/resources-in-the-rest-api#user-agent-required
+        .header(reqwest::header::USER_AGENT, USER_AGENT);
+    if let Some(etag) = etag {
+        request = request.header(reqwest::header::IF_NONE_MATCH, etag);
+    }
+
+    let result = match request.send().await {
+        Ok(r) => r,
+        Err(e) => return UpdateResult::RewestError(e),
+    };
+    match result.status() {
+        StatusCode::NOT_MODIFIED => UpdateResult::NotModified,
+        StatusCode::OK => {
+            #[derive(serde::Deserialize)]
+            struct JsonRelease {
+                name: String,
+            }
+
+            let etag = result
+                .headers()
+                .get(reqwest::header::ETAG)
+                // Its technically allowed to have non-valid ascii data in the header,
+                // in which case we can't do anything except set etag=None.
+                .map(|h| h.to_str())
+                .transpose()
+                .unwrap_or_default()
+                .map(|s| s.to_string());
+
+            match result.json::<JsonRelease>().await {
+                Ok(r) => UpdateResult::Update(Release {
+                    version: r.name,
+                    etag,
+                }),
+                Err(e) => UpdateResult::RewestError(e),
+            }
+        }
+        other => UpdateResult::Other(anyhow::anyhow!(
+            "Unexpected response status code: {}",
+            other
+        )),
```

Comment:
I guess github might pull some temporary limits on the api ... But I guess we'll see when this happens the next time. Most of them will probably be just ReqwestErrors due to timeouts.

Reply:
Yeah there is rate limiting, but currently it looks to be about 5000 per day -- but they do mention this shouldn't be taken as a given and ToS can deny you at any point. But I think 10min poll rate is `6x24=144` requests a day which should be okay.

Reply:
I was thinking of like critical failures which have been happening every now and then. The etag If-None-Match did not seem to be counted towards ratelimits told back to the user, so under normal ops this should be very much polite enough.